source/swap/contracts/Swap.sol

```diff
@@ -522,24 +517,21 @@ contract Swap is ISwap, Ownable {
    * @notice Marks a nonce as used for the given signer
    * @param signer address Address of the signer for which to mark the nonce as used
    * @param nonce uint256 Nonce to be marked as used
-   * @return bool True if the nonce was not marked as used already
    */
   function _markNonceAsUsed(address signer, uint256 nonce)
     internal
-    returns (bool)
   {
     uint256 groupKey = nonce / 256;
     uint256 indexInGroup = nonce % 256;
     uint256 group = _nonceGroups[signer][groupKey];
 
-    // If it is already used, return false
+    // If it is already used, return mit cancel and revert
     if ((group >> indexInGroup) & 1 == 1) {
-      return false;
+      emit Cancel(nonce, msg.sender);
```

Comment:
`Cancel` should only be emitted from the `cancel` function. Would place it after L319.

Reply:
Makes ssenes